vault/mfa/totp.go

```diff
@@ -0,0 +1,127 @@
+package mfa
+
+import (
+	"bytes"
+	"encoding/base64"
+	"fmt"
+	"image/png"
+	"time"
+
+	"github.com/hashicorp/vault/logical"
+	"github.com/pquerna/otp/totp"
+)
+
+// createTOTPKey uses the given method and identifier to create a TOTP secret.
+// It returns an error on error, but may also add warnings to the response
+// object.
+func (b *backend) createTOTPKey(method *mfaMethodEntry, identifierEntry *mfaIdentifierEntry, resp *logical.Response) error {
+	if identifierEntry.TOTPAccountName == "" {
+		newResp := logical.ErrorResponse("\"totp_account_name\" must be set on the identifier")
+		*resp = *newResp
+		return nil
+	}
+
+	alg, err := method.totpAlgorithm()
+	if err != nil {
+		return err
+	}
+
+	opts := totp.GenerateOpts{
+		Algorithm:   alg,
+		AccountName: identifierEntry.TOTPAccountName,
+		Issuer:      fmt.Sprintf("Vault MFA: %s/%s", method.Name, identifierEntry.Identifier),
+	}
+
+	if opts.AccountName != "" {
+		opts.Issuer = fmt.Sprintf("%s (%s)", opts.Issuer, opts.AccountName)
+	}
+
+	key, err := totp.Generate(opts)
+	if err != nil {
+		return err
+	}
+	if key == nil {
+		return fmt.Errorf("generated TOTP key is nil")
+	}
+
+	// We store the original URL so we have the full set of parameters in the
+	// key, but also the secret itself as a shortcut to avoid having to parse
+	// the URL over and over
+	identifierEntry.TOTPURL = key.String()
+	identifierEntry.TOTPSecret = key.Secret()
+
+	if resp.Data == nil {
+		resp.Data = map[string]interface{}{}
+	}
+
+	resp.Data["totp_secret"] = key.Secret()
+	resp.Data["totp_url"] = key.String()
+
+	// Generate a PNG. This isn't super huge and can be quite helpful to the
+	// end user.
+	keyImage, err := key.Image(1024, 1024)
```

Comment:
I wouldn't include this in the API response. Seems like something that should be done client side, not part of the server payload.